src/services/TestResultsService.ts

```diff
@@ -1,247 +1,157 @@
-          if (!newTestResult.testHistory) {
-            newTestResult.testHistory = [oldTestResult];
-          } else {
-            delete oldTestResult.testHistory;
-            newTestResult.testHistory.push(oldTestResult);
-          }
-          return this.testResultsDAO.updateTestResult(newTestResult)
-              .then((data) => {
-                return newTestResult;
-              }).catch((error) => {
-                throw new HTTPError(500, error.message);
-              });
-        }).catch((error) => {
-          throw new HTTPError(error.statusCode, error.body);
-        });
-  }
-
-  private arrayCustomizer(objValue: any, srcValue: any) {
-    if (isArray(objValue) && isArray(srcValue)) {
-      return srcValue;
+    try {
+      const result = await this.testResultsDAO.getBySystemNumber(systemNumber);
+      const response: models.ITestResultData = { Count: result.Count, Items: result.Items };
+      const testResults = ValidationUtil.checkTestResults(response);
+      const oldTestResult = this.getTestResultToArchive(testResults, payload.testResultId);
+      oldTestResult.testVersion = enums.TEST_VERSION.ARCHIVED;
+      const newTestResult: models.ITestResult = cloneDeep(oldTestResult);
+      newTestResult.testVersion = enums.TEST_VERSION.CURRENT;
+      mergeWith(newTestResult, payload, MappingUtil.arrayCustomizer);
+      if (this.shouldGenerateNewTestCodeRe(oldTestResult, newTestResult)) {
+        const vehicleSubclass = newTestResult.vehicleSubclass && newTestResult.vehicleSubclass.length ? newTestResult.vehicleSubclass[0] : undefined;
+        await this.getTestTypesWithTestCodesAndClassification((newTestResult.testTypes as any[]),
+          newTestResult.vehicleType, newTestResult.vehicleSize, newTestResult.vehicleConfiguration,
+          newTestResult.noOfAxles, newTestResult.euVehicleCategory, newTestResult.vehicleClass.code,
+          vehicleSubclass, newTestResult.numberOfWheelsDriven);
+      }
+      await this.checkTestTypeStartAndEndTimestamp(oldTestResult, newTestResult);
+      MappingUtil.manageDefectsArray(newTestResult);
+      MappingUtil.setAuditDetails(newTestResult, oldTestResult, msUserDetails);
+      if (!newTestResult.testHistory) {
+        newTestResult.testHistory = [oldTestResult];
+      } else {
+        delete oldTestResult.testHistory;
+        newTestResult.testHistory.push(oldTestResult);
+      }
+      try {
+        await this.testResultsDAO.updateTestResult(newTestResult);
+        return newTestResult;
+      } catch (dynamoError) {
+        throw new models.HTTPError(500, dynamoError.message);
+      }
+    } catch (error) {
+      throw new models.HTTPError(error.statusCode, error.body);
     }
   }
 
-  public shouldGenerateNewTestCodeRe(oldTestResult: ITestResult, newTestResult: ITestResult) {
+  public shouldGenerateNewTestCodeRe(oldTestResult: models.ITestResult, newTestResult: models.ITestResult) {
```

Comment:
Need to rename the function

Reply:
This chunk of code may get refactored in 18783. Renaming it would not be of benefit at the moment.